mrmustard/lab_dev/samplers.py

```diff
@@ -0,0 +1,229 @@
+        return rng.choice(
+            a=list(product(self.meas_outcomes, repeat=len(state.modes))),
+            p=self.probabilities(state),
+            size=n_samples,
+        )
+
+    def _get_povm(self, meas_outcome: Any, mode: int) -> CircuitComponent:
+        r"""
+        Returns the POVM associated with a given outcome on a given mode.
+
+        Args:
+            meas_outcome: The measurement outcome.
+            mode: The mode.
+
+        Returns:
+            The POVM circuit component.
+        """
+        if isinstance(self.povms, CircuitComponent):
+            kwargs = self.povms.parameter_set.to_dict()
+            kwargs[self._outcome_arg] = meas_outcome
+            return self.povms.__class__(modes=[mode], **kwargs)
+        else:
+            return self.povms[self.meas_outcomes.index(meas_outcome)].on([mode])
+
+    def _validate_probs(self, probs: Sequence[float], atol: float) -> Sequence[float]:
+        r"""
+        Validates that the given probability distribution sums to ``1`` within some
+        tolerance and returns a renormalized probability distribution to account for
+        small numerical errors.
+
+        Args:
+            probs: The probability distribution to validate.
+            atol: The absolute tolerance to validate with.
+        """
+        atol = atol or settings.ATOL
+        prob_sum = sum(probs)
+        if not math.allclose(prob_sum, 1, atol):
+            raise ValueError(f"Probabilities sum to {prob_sum} and not 1.0.")
+        return math.real(probs / prob_sum)
+
+
+class PNRSampler(Sampler):
+    r"""
+    A sampler for photon-number resolving (PNR) detectors.
+
+    Args:
+        cutoff: The photon number cutoff.
+    """
+
+    def __init__(self, cutoff: int) -> None:
+        super().__init__(list(range(cutoff)), Number([0], 0))
+        self._cutoff = cutoff
+        self._outcome_arg = "n"
+
+    def probabilities(self, state, atol=1e-4):
+        return self._validate_probs(state.fock_distribution(self._cutoff), atol)
+
+
+class HomodyneSampler(Sampler):
+    r"""
+    A sampler for homodyne measurements.
+
+    Args:
+        phi: The quadrature angle where ``0`` corresponds to ``x`` and ``\pi/2`` to ``p``.
+        bounds: The range of values to discretize over.
+        num: The number of points to discretize over.
+    """
+
+    def __init__(
+        self,
+        phi: float = 0,
+        bounds: tuple[float, float] = (-10, 10),
+        num: int = 1000,
+    ) -> None:
+        meas_outcomes, step = np.linspace(*bounds, num, retstep=True)
+        super().__init__(
+            list(meas_outcomes),
+            QuadratureEigenstate([0], x=0, phi=phi),
+        )
+        self._step = step
+        self._outcome_arg = "x"
+
+    def probabilities(self, state, atol=1e-4):
+        probs = state.quadrature_distribution(
+            self.meas_outcomes, self.povms.phi.value[0]
+        ) * self._step ** len(state.modes)
```

Comment:
I see here these are the probabilities related to measuring all modes, but since we are measuring one mode at a time, it may be simpler than this.

Reply:
Yes, our sample method does for one mode but I think this is a convenient way for users to compute all modes if they want to